Source/DafnyStandardLibraries/src/DafnyStdLibs/DynamicArray.dfy

```diff
@@ -0,0 +1,244 @@
+module DafnyStdLibs.DynamicArray {
+  import opened BoundedInts
+  import opened Wrappers
+
+  /**
+  The `DynamicArray` module and class define a data structure that has the same performance characteristics of an array, 
+  but additionally allows growing and shrinking the size of the array. The cost of shrinking the array is constant time, 
+  while the cost of growing it by a single element is either constant time or linear in the size of the array. 
+  When growing the array multiple times, the amortized cost of growing the array by one element is constant time.
+
+  The downsize of using a DynamicArray is that it occupies up to twice the amount of memory as its size, 
+  and when elements are removed, its occupied memory does not decrease. 
+  */
+  class DynamicArray<A> {
+    ghost var items : seq<A>
+    ghost var Repr : set<object>
+
+    const defaultValue: A
+    var size: uint32
+    var capacity: uint32
+    var data: array<A>
+
+    const MAX_CAPACITY: uint32 := (TWO_TO_THE_32 - 1) as uint32
+    const MAX_CAPACITY_BEFORE_DOUBLING: uint32 := MAX_CAPACITY / 2
```

Comment:
I agree with not modeling the max capacity and just assuming it always succeeds - since the runtime error when failing to allocate the storage will be unrecoverable in Dafny source code, it's not a soundness risk. If anything making client code transitively handle the failure makes it more likely that a program will recover from running out of memory and limp along when it shouldn't.

It's fine for now if the JSON library this came from continues this design internally when we import it over, though, especially since it already has to model runtime failure for various other reasons anyway: https://github.com/dafny-lang/libraries/blob/ae8708c091d32383235d5d8c15c08cff05613bbc/src/JSON/Errors.dfy